source/MaterialXShaderGen/ShaderGenerators/Glsl/GlslShaderGenerator.cpp

```diff
@@ -339,10 +363,45 @@ void GlslShaderGenerator::emitFunctionDefinitions(Shader& shader)
     _bsdfNodeArguments[0] = Argument("vec3", "wi");
     _bsdfNodeArguments[1] = Argument("vec3", "wo");
 
-    // Emit function for handling texture coords v-flip 
-    // as needed by the v-direction set by the user
     BEGIN_SHADER_STAGE(shader, HwShader::PIXEL_STAGE)
+
+        // Emit function for handling texture coords v-flip 
+        // as needed by the v-direction set by the user
         shader.addBlock(shader.getRequestedVDirection() != getTargetVDirection() ? VDIRECTION_FLIP : VDIRECTION_NOOP);
+
+        // For surface shaders we need light shaders
+        if (shader.hasClassification(SgNode::Classification::SHADER | SgNode::Classification::SURFACE))
+        {
+            // Emit functions for all bound light shaders
+            for (auto lightShader : getBoundLightShaders())
+            {
+                lightShader.second->emitFunctionDefinition(SgNode::NONE, *this, shader);
+            }
+
+            // Emit active light count function
+            shader.addLine("int numActiveLightSources()", false);
+            shader.beginScope(Shader::Brackets::BRACES);
+            shader.addLine("return min(u_numActiveLightSources, MAX_LIGHT_SOURCES)");
```

Comment:
So this allows for a block size of "max" light sources but a clamp of only having N of them active?
Is this so that we pre-allocate the memory in the shader and can dynamically change the number of active lights up to the max without recompiling the shader?

Reply:
But if we change the max light count we do need to recompile since it's a #define.

Reply:
Yes, this could be my limited knowledge of GLSL.. Is it possible to declare this as a variable size array, and have the size set by application when binding light sources? That would be better if it works.